Applications/Spire/Source/Ui/ContextMenu.cpp

```diff
@@ -0,0 +1,334 @@
+#include "Spire/Ui/ContextMenu.hpp"
+#include <QCoreApplication>
+#include <QHBoxLayout>
+#include <QKeyEvent>
+#include <QScreen>
+#include <QTimer>
+#include "Spire/Spire/ArrayListModel.hpp"
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Ui/ListItem.hpp"
+#include "Spire/Ui/ListView.hpp"
+#include "Spire/Ui/OverlayPanel.hpp"
+#include "Spire/Ui/SubMenuItem.hpp"
+#include "Spire/Ui/TextBox.hpp"
+
+using namespace boost;
+using namespace boost::signals2;
+using namespace Spire;
+using namespace Styles;
+
+namespace {
+  const auto MAX_WIDTH = 130;
+  const auto MARGIN_SIZE = 50;
+
+  auto MARGIN_HEIGHT() {
+    static auto height = scale_height(MARGIN_SIZE);
+    return height;
+  }
+
+  auto MARGIN_WIDTH() {
+    static auto width = scale_width(MARGIN_SIZE);
+    return width;
+  }
+
+  auto LIST_VIEW_STYLE(StyleSheet style) {
+    style.get(Any()).
+      set(EdgeNavigation::CONTAIN).
+      set(SelectionMode::NONE);
+    return style;
+  }
+}
+
+ContextMenu::ContextMenu(QWidget& parent)
+    : QWidget(&parent),
+      m_active_menu_window(nullptr) {
+  setAttribute(Qt::WA_Hover);
+  setMaximumWidth(scale_width(MAX_WIDTH));
+  m_list = std::make_shared<ArrayListModel<MenuItem>>();
+  m_list_view = new ListView(m_list,
+    std::bind_front(&ContextMenu::build_item, this));
+  m_list_view->setSizePolicy(QSizePolicy::Preferred, QSizePolicy::Expanding);
+  update_style(*m_list_view, [&] (auto& style) {
+    style = LIST_VIEW_STYLE(style);
+  });
+  m_list_view->connect_submit_signal(
+    std::bind_front(&ContextMenu::on_submit, this));
+  m_list_view->installEventFilter(this);
+  setFocusProxy(m_list_view);
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins({});
+  layout->addWidget(m_list_view);
+  m_window = new OverlayPanel(*this, parent);
+  m_window->setWindowFlags(Qt::Popup | (m_window->windowFlags() & ~Qt::Tool));
+  m_window->set_closed_on_focus_out(true);
+  m_window->set_positioning(OverlayPanel::Positioning::NONE);
+  m_window->set_is_draggable(false);
+  m_window->setMouseTracking(true);
+  m_window->installEventFilter(this);
+  m_window->parentWidget()->installEventFilter(this);
+  on_window_style();
+  m_window_style_connection =
+    connect_style_signal(*m_window, [=] { on_window_style(); });
+}
+
+void ContextMenu::add_menu(const QString& name, const ContextMenu& menu) {
+  menu.window()->setParent(this);
+  m_list->push(MenuItem{MenuItemType::SUBMENU, name, &menu});
+}
+
+void ContextMenu::add_action(const QString& name, const Action& action) {
+  m_list->push(MenuItem{MenuItemType::ACTION, name, action});
+}
+
+std::shared_ptr<BooleanModel> ContextMenu::add_check_box(const QString& name) {
+  auto model = std::make_shared<LocalBooleanModel>();
```

Comment:
Avoid duplicating implementations, when possible implement one function in terms of another.